rspace/src/main/scala/coop/rchain/rspace/RSpace.scala

```diff
@@ -223,107 +222,84 @@ class RSpace[F[_], C, P, A, K](
              * In this version, we also add the produced data directly to this cache.
              */
             channelToIndexedDataList <- channels.traverse { c: C =>
-                                         (for {
-                                           data <- store.getData(c)
-                                         } yield (Random.shuffle(data.zipWithIndex)))
-                                           .map(
-                                             as =>
-                                               c -> {
-                                                 if (c == batChannel)
-                                                   (data, -1) +: as
-                                                 else as
-                                               }
-                                           )
+                                         for {
+                                           d  <- store.getData(c)
+                                           as = Random.shuffle(d.zipWithIndex)
+                                           sp = if (c == batChannel) (data, -1) +: as else as
+                                         } yield (c -> sp)
                                        }
             firstMatch <- extractFirstMatch(
                            channels,
                            matchCandidates,
                            channelToIndexedDataList.toMap
                          )
-          } yield firstMatch match {
-            case None             => remaining.asLeft[MaybeProduceCandidate]
-            case produceCandidate => produceCandidate.asRight[Seq[CandidateChannels]]
-          }
+          } yield
+            firstMatch match {
+              case None             => remaining.asLeft[MaybeProduceCandidate]
+              case produceCandidate => produceCandidate.asRight[Seq[CandidateChannels]]
+            }
       }
     groupedChannels.tailRecM(go)
   }
 
   private[this] def processMatchFound(
       pc: ProduceCandidate[C, P, A, K]
-  ): F[MaybeActionResult] =
-    pc match {
-      case ProduceCandidate(
-          channels,
-          WaitingContinuation(
-            patterns,
+  ): F[MaybeActionResult] = {
+    val ProduceCandidate(
+      channels,
+      WaitingContinuation(
+        patterns,
+        continuation,
+        persistK,
+        peeks,
+        consumeRef
+      ),
+      continuationIndex,
+      dataCandidates
+    ) = pc
+
+    def removeMatchedDatumAndJoin(channelsToIndex: Map[C, Int]): F[Seq[Unit]] =
+      dataCandidates
+        .sortBy(_.datumIndex)(Ordering[Int].reverse)
```

Comment:
Why does it have to be sorted?

Reply:
Looks like we're traversing it wholly anyway and we're interested only in `dataIndex >= 0`